File: app/src/main/java/com/jbproductions/wmbr/XmlParser.java
package com.jbproductions.wmbr;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;
import org.xmlpull.v1.XmlPullParserFactory;
import java.io.IOException;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

public class XmlParser {

    public XmlParser() {

    }

    private static final String CURRENT_INFO_URL = "https://wmbr.org/cgi-bin/xmlinfo";
    private static final String TRACK_BLASTER_URL = "https://www.track-blaster.com/wmbr/pl_recent_songs.php";


    /**
     * Pulls in current information (show, host, weather, etc) from WMBR's xml API.
     * @return HashMap with values mapped to their tags as seen at 'https://wmbr.org/cgi-bin/xmlinfo'
     */
    static Map getStreamMetadata() {
        Map<String, String> wmbrStatusMap = new HashMap<>();
        XmlPullParser parser = setupXmlParser(CURRENT_INFO_URL);

        try {
            int eventType = parser.getEventType();

            while(eventType != XmlPullParser.END_DOCUMENT) {
                String tagName;

                if(eventType == XmlPullParser.START_TAG) {

                    tagName = parser.getName();

                    if ("time".equals(tagName)) {
                        wmbrStatusMap.put("time", parser.nextText());
                    } else if ("showname".equals(tagName)) {
                        // If the showname has an irregular character, it may be parsed incorrectly due to mismatched encodings
                        // This block will attempt to convert to standard UTF-8
                        String utfString;
                        try {
                            utfString = new String(parser.nextText().getBytes(StandardCharsets.ISO_8859_1), StandardCharsets.UTF_8);
                            wmbrStatusMap.put("showName", utfString);
                        } catch (UnsupportedEncodingException e) {
                            wmbrStatusMap.put("showName", parser.nextText());
                            e.printStackTrace();
                        }
                    } else if ("showhosts".equals(tagName)) {
                        // If the host name has an irregular character, it may be parsed incorrectly due to mismatched encodings
                        // This block will attempt to convert to standard UTF-8
                        String utfString;
                        try {
                            utfString = new String(parser.nextText().getBytes(StandardCharsets.ISO_8859_1), StandardCharsets.UTF_8);
                            wmbrStatusMap.put("showHosts", utfString);
                        } catch (UnsupportedEncodingException e) {
                            wmbrStatusMap.put("showHosts", parser.nextText());
                            e.printStackTrace();
                        }
                    } else if ("showid".equals(tagName)) {
                        wmbrStatusMap.put("showID", parser.nextText());
                    } else if ("showurl".equals(tagName)) {
                        wmbrStatusMap.put("showUrl", parser.nextText());
                    } else if ("temp".equals(tagName)) {
                        /* Android interprets the degree symbol incorrectly because of its default text encoding.
                        To get it to display properly, we pass the text bitstream to a new UTF-8 string */
                        String utfString;
                        try {
                            utfString = new String(parser.nextText().getBytes(StandardCharsets.ISO_8859_1), StandardCharsets.UTF_8);
                            wmbrStatusMap.put("temperature", utfString);
                        } catch (UnsupportedEncodingException e) {
                            wmbrStatusMap.put("temperature", parser.nextText());
                            e.printStackTrace();
                        }
                    } else if ("wx".equals(tagName)) {
                        wmbrStatusMap.put("wx", parser.nextText());
                    }
                }

                eventType = parser.next();  // Advance the parser to the next tag
            }

        } catch (XmlPullParserException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return wmbrStatusMap;

    }

    /**
     * Configures an XmlPullParser object to pull xml data from a given URL
     * @param resourceURL String of the URL to the xml that will be parsed
     * @return XmlPullParser object that can be stepped through and parsed
     */
    static XmlPullParser setupXmlParser(String resourceURL) {
        XmlPullParserFactory parserFactory;
        try {
            parserFactory = XmlPullParserFactory.newInstance();
            XmlPullParser parser = parserFactory.newPullParser();
            InputStream xmlInputStream = downloadUrl(resourceURL);
            parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false);
            parser.setInput(xmlInputStream, null);
            return parser;
        } catch (XmlPullParserException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * Given a string representation of a URL, sets up a connection and gets an input stream.
     * @param urlString String representation of URL for XML file
     * @return InputStream
     * @throws IOException when connection fails
     */
    private static InputStream downloadUrl(String urlString) throws IOException {
        URL url = new URL(urlString);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setReadTimeout(10000 /* milliseconds */);
        conn.setConnectTimeout(15000 /* milliseconds */);
        conn.setRequestMethod("GET");
        conn.setDoInput(true);
        // Starts the query
        conn.connect();
        return conn.getInputStream();
    }
}

File: app/src/main/java/com/jbproductions/wmbr/util/XmlParser.kt
package com.jbproductions.wmbr.util

import android.util.Xml
import com.jbproductions.wmbr.model.Show
import com.jbproductions.wmbr.model.StreamInfo
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL

private const val INFO_URL : String = "https://wmbr.org/cgi-bin/xmlinfo"
private const val SCHEDULE_URL : String = "https://wmbr.org/cgi-bin/xmlsched"
//private const val TRACK_BLASTER_URL : String = "https://www.track-blaster.com/wmbr/pl_recent_songs.php"
private val ns : String? = null

@Throws(XmlPullParserException:: class, IOException::class)
fun parseStreamInfo() : StreamInfo {

    val inputStream = getInputStream(INFO_URL)

    inputStream.use {
        val parser : XmlPullParser = Xml.newPullParser()
        parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
        parser.setInput(inputStream, null)
        parser.nextTag()
        return getInfo(parser)
    }
}

@Throws(XmlPullParserException:: class, IOException::class)
fun parseSchedule() : List<*> {

    val inputStream = getInputStream(SCHEDULE_URL)

    inputStream.use {
        val parser : XmlPullParser = Xml.newPullParser()
        parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
        parser.setInput(inputStream, null)
        parser.nextTag()
        return buildShowList(parser)
    }
}

// Given a string representation of a URL, sets up a connection and gets
// an input stream.
@Throws(IOException::class)
private fun getInputStream(urlString: String): InputStream? {
    val url = URL(urlString)
    return (url.openConnection() as? HttpURLConnection)?.run {
        readTimeout = 10000
        connectTimeout = 15000
        requestMethod = "GET"
        doInput = true
        // Starts the query
        connect()
        inputStream
    }
}

@Throws(XmlPullParserException::class, IOException::class)
private fun getInfo(parser: XmlPullParser) : StreamInfo {

    parser.require(XmlPullParser.START_TAG, ns, "wmbrinfo")
    val streamInfo = StreamInfo()

    while (parser.next() != XmlPullParser.END_TAG) {
        if (parser.eventType != XmlPullParser.START_TAG) {
            continue
        }
        when (parser.name) {
            "time" -> streamInfo.time = readString(parser, "time")
            "showname" -> streamInfo.showName = readString(parser, "showname")
            "showhosts" -> streamInfo.hosts = readString(parser, "showhosts")
            "showid" -> streamInfo.id = readInt(parser, "showid")
            "showurl" -> streamInfo.url = readString(parser, "showurl")
            else -> skip(parser)
        }
    }

    return streamInfo
}

@Throws(XmlPullParserException::class, IOException::class)
private fun buildShowList(parser: XmlPullParser) : List<Show> {

    parser.require(XmlPullParser.START_TAG, ns, "wmbr_schedule")
    val shows = mutableListOf<Show>()

    while (parser.next() != XmlPullParser.END_TAG) {
        if (parser.eventType != XmlPullParser.START_TAG) {
            continue
        }
        // Starts by looking for the entry tag
        if (parser.name == "show") {
            shows.add(readShow(parser))
        } else {
            skip(parser)
        }
    }

    return shows
}

// Parses the contents of a show. If it encounters a valid tag, hands them off
// to their respective "read" methods for processing. Otherwise, skips the tag.
@Throws(XmlPullParserException::class, IOException::class)
private fun readShow(parser: XmlPullParser) : Show {

    parser.require(XmlPullParser.START_TAG, ns, "show")
    val newShow = Show()

    while (parser.next() != XmlPullParser.END_TAG) {
        if (parser.eventType != XmlPullParser.START_TAG) {
            continue
        }
        when (parser.name) {
            "id" -> newShow.id = readInt(parser, "id")
            "name" -> newShow.name = readString(parser, "name")
            "day" -> newShow.day = readInt(parser, "day")
            "time" -> newShow.time = readInt(parser, "time")
            "length" -> newShow.length = readInt(parser, "length")
            "hosts" -> newShow.hosts = readString(parser, "hosts")
            "description" -> newShow.description = readString(parser, "description")
            "url" -> newShow.url = readString(parser, "url")
            "email" -> newShow.emailPrefix = readString(parser, "email")
            "alternates" -> newShow.alternates = readInt(parser, "alternates") != 0
            else -> skip(parser)
        }
    }

    return newShow
}

// Processes name tags in the feed.
@Throws(IOException::class, XmlPullParserException::class)
private fun readString(parser: XmlPullParser, tag: String) : String {
    parser.require(XmlPullParser.START_TAG, ns, tag)
    val name = readText(parser)
    parser.require(XmlPullParser.END_TAG, ns, tag)
    return name
}

@Throws(IOException::class, XmlPullParserException::class)
private fun readInt(parser: XmlPullParser, tag: String) : Int {
    parser.require(XmlPullParser.START_TAG, ns, tag)
    val day = readText(parser).toInt()
    parser.require(XmlPullParser.END_TAG, ns, tag)
    return day
}

@Throws(IOException::class, XmlPullParserException::class)
private fun readText(parser : XmlPullParser) : String {
    var result =""
    if (parser.next() == XmlPullParser.TEXT) {
        result = parser.text
        parser.nextTag()
    }
    return result
}

@Throws(IOException::class, XmlPullParserException::class)
private fun skip(parser : XmlPullParser) {
    if (parser.eventType != XmlPullParser.START_TAG) {
        throw IllegalStateException()
    }
    var depth = 1
    while (depth != 0) {
        when (parser.next()) {
            XmlPullParser.END_TAG -> depth--
            XmlPullParser.START_TAG -> depth++
        }
    }
}